### DISCUSSION TITLE
how the next-js is working under the hood

### DISCUSSION THREAD

**Discussion by Zain-ul-din** [vercel/next.js#44004] [Help] [answered]
Hey, I'm a student I know this question is a little bit dumb but I also wish to know how the next js is importing all code that is exported by us (users). for example:-

```js
  // in react-js it makes a lot more sense
  root.render (<>My Components </>)

 // how does next-js know about my exports?
 export default function App () { 
    return (<>Home</>)
 }
```
I want to implement and learn these features on my own please also suggest me road map that I can follow.

**Thanks in Advance!**

**Comment by icyJoseph** [comment, marked as answer]
When you have default exports you can target the `default` key of the module, and use that to render. They go through every file in pages (except api routes, for React rendering that is) and collect the default exports, as well as a couple of well known named exports such as getStaticProps or getServerSideProps.

So roughly speaking:

- walk the pages directory
- import each path your find
- read the default property from the import, as well as other valuable named exports
- render to string using [react-dom/server](https://reactjs.org/docs/react-dom-server.html)
- generate html pages

Then you need to write a router that handles SPA-like transitions.

You also need a server to render on-demand requests.

Now that `appDir` is in the works it is slightly more difficult to follow the flow but these files should guide you around:

- https://github.com/vercel/next.js/blob/canary/packages/next/server/app-render.tsx
- https://github.com/vercel/next.js/blob/canary/packages/next/export/worker.ts
- https://github.com/vercel/next.js/blob/canary/packages/next/build/index.ts
- https://github.com/vercel/next.js/blob/canary/packages/next/server/render.tsx
- https://github.com/vercel/next.js/blob/canary/packages/next/server/load-components.ts

And the question is not dumb at all :) this is a tricky problem to solve, specially because Next.js pre-renders pages, while keeping others on-demand, and still can offer SPA transitions! 

You can try this without Next.js in the mix. Make a codesandbox CRA project and put this inside the `index.js` file:

```js
import { StrictMode } from "react";
import { createRoot } from "react-dom/client";
import { renderToString } from 'react-dom/server' // -> add this part

import App from "./App";

const rootElement = document.getElementById("root");
const root = createRoot(rootElement);

root.render(
  <StrictMode>
    <App />
  </StrictMode>
);

console.log(renderToString(<App/>)) // -> add this part too
```

**Reply by Zain-ul-din** [reply]
@icyJoseph thanks for your answer and kindness.
please can you tell me what library ```next js``` is used to track each file in the page directory?

**Reply by icyJoseph** [reply]
I think for that you can take a look here: https://github.com/vercel/next.js/blob/canary/packages/next/build/entries.ts
